src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java

```diff
@@ -467,6 +450,36 @@ public Repository getOrCheckoutRepository(VcsRepositoryUrl sourceRepoUrl, VcsRep
         }
     }
 
+    /**
+     * Waits until no clone operation is running for the given path.
+     *
+     * Retries once a second for up to {@link #JGIT_TIMEOUT_IN_SECONDS} seconds before giving up.
+     *
+     * @param localPath The path in which a clone operation should be made.
+     * @throws CanceledException If the waiting has been interrupted.
+     * @throws GitException If the path is still busy after the maximum number of retries.
+     */
+    private void waitUntilPathNotBusy(final Path localPath) throws CanceledException, GitException {
+        int remainingSeconds = JGIT_TIMEOUT_IN_SECONDS;
+
+        while (cloneInProgressOperations.containsKey(localPath)) {
+            log.warn("Clone is already in progress. This will lead to an error. Wait for a second");
+            try {
+                Thread.sleep(1000);
+            }
+            catch (InterruptedException ex) {
+                throw new CanceledException("Waiting for local path to be free for cloning got interrupted.");
+            }
+
+            if (remainingSeconds == 0) {
```

Comment:
```suggestion
            if (remainingSeconds <= 0) {
```

One minor suggestion.